### DISCUSSION TITLE
Usability: command-line options for node configuration

### DISCUSSION THREAD

**Discussion by mzabaluev** [movementlabsxyz/movement#385] [Ideas]
Tracking issue: #481 

Currently, node configuration is obtained from environment variables.
A system based on configuration files is discussed in #384, but we can also consider making some node settings configurable via command-line options.

**Comment by l-monninger** [comment]
I think we should use the following hierarchy in pseudo-JavaScript:

```typescript
cliArgument ?? configFileField ?? environmentVariable ?? defaultValue;
```

**Reply by l-monninger** [reply]
Atlantis is a good example of this: https://www.runatlantis.io/docs/server-configuration

**Reply by mzabaluev** [reply]
Yes, this is pretty conventional, for example git follows this hierarchy.

**Reply by l-monninger** [reply]
I believe this is ready for an issue.

**Comment by mzabaluev** [comment]
Should we come up with a list of options that make sense to be overridable in the command line?
I'd like to identify use cases that are important to:

1. Node operators.
2. Developers of tests.

A less customer-driven approach can be also taken: anything that does not have a sensible default should be settable with a CLI option (overriding the other configuration sources listed in [another comment](#discussioncomment-10394002)).

**Reply by l-monninger** [reply]
I agree, these are the main cases. 

> anything that does not have a sensible default 

I kind of think we should strive to design our systems s.t. there is always a sensible default--at least assuming a particular running mode is applied elsewhere.

**Comment by mzabaluev** [comment]
Proposed options for suzuka-full-node:

* Chain ID (as `--chain-id`).
* DA light node URL (as `--da-light-node`).
* Location of the secrets file providing e.g. the executor private key (as `--secrets-file`). See the discussion in #387 for other options.

**Reply by l-monninger** [reply]
> Location of the executor private key (as --secrets-file).

This will not be so simple because of considerations in #435 . Really, we will probably want some string format or multiple fields to represent how a structure for signing can be deserialized from CLI args.

**Reply by l-monninger** [reply]
Are these supposed to be additional options from what is currently provided? I would say we kind of want everything under the current config until we figure out how we want to strip out. For example, a node operator may want to play around with load-shedding parameters and find it easier to do that playing around with the CLI. x

**Reply by mzabaluev** [reply]
It's difficult to make assumptions without customer input.

My current thinking is to list the options that need to be customized for any deployment, and provide a way to specify locations of some important files.

Perhaps the dot-movement directory path should be a CLI option, too. But it's more natural to think of it as the shared environment for possibly multiple local processes, and it has a well-known default, so it could be a common convention to set it as an environment variable.

**Reply by mzabaluev** [reply]
> > Location of the executor private key (as --secrets-file).
> 
> This will not be so simple because of considerations in #435 . Really, we will probably want some string format or multiple fields to represent how a structure for signing can be deserialized from CLI args.

Yes, I really mean the secrets file should serialize a structure configuring access to the private keys.
In the simplest form, for purposes of local testing, it will encode a single private key.

**Reply by mzabaluev** [reply]
Added `--da-light-node` URL, a moving part in the two-service configuration.

**Comment by mzabaluev** [comment]
Proposed options for m1-da-light-node:

* Chain ID as `--chain-id`.
* Celestia RPC URL as `--celestia-rpc`.
* Celestia Websocket URL as `--celestia-websocket`.
* ~Migrations configuration file (see #470) as `--da-migrations`.~

**Reply by l-monninger** [reply]
Yeah, so, currently, I adopted the approach of requesting protocol, hostname, and port separately. But, do we want to make these all `Url` parsed?

**Reply by mzabaluev** [reply]
For configuring clients of HTTP-based protocols it's more convenient to use URLs, especially in environment variables.

For local service configuration, we could do either, but I think the string representation of `SocketAddr` is a widely adopted microformat so we can change to it for convenience.

**Reply by mzabaluev** [reply]
The migrations could be a well-known file under dot-movement, too.

**Reply by l-monninger** [reply]
Yes, ultimately, that's how I envision it actually be stored as a file.

**Reply by l-monninger** [reply]
One thing that we still need to break out is how we want config files to be stored. That is, do we want multiple configs, a single config with a global namespace, etc. There's a single-source of truth vs. modularity tradeoff here.

**Reply by mzabaluev** [reply]
> One thing that we still need to break out is how we want config files to be stored. That is, do we want multiple configs, a single config with a global namespace, etc. There's a single-source of truth vs. modularity tradeoff here.

Yes, this should be discussed on the side of #384.
@l-monninger Should the config location (modulo the decision on whether it's global or service-specific config) be customizable at least? Dot-movement is an easy default solution, but service deployments and filesystem standards on Unix preferred to split configuration from dynamic storage for decades now. With modern immutable container configurations, having the config deployed in a read-only `/etc` layer may be a requirement.

**Reply by mzabaluev** [reply]
Broken out as #533.
